File: algorithms/cpp/randomFlipMatrix/randomFlipMatrix.cpp
// Source: https://leetcode.com/problems/random-flip-matrix/
// Author: Miao Zhang
// Date:   2021-02-19

class Solution {
public:
    Solution(int n_rows, int n_cols): rows_(n_rows),
        cols_(n_cols), n_(n_rows * n_cols) {

    }

    vector<int> flip() {
        int s = rand() % (n_--);
        int index = s;
        if (dic_.count(s)) {
            index = dic_[s];
        }
        dic_[s] = dic_.count(n_) ? dic_[n_] : n_;
        return {index / cols_, index % cols_};
    }

    void reset() {
        dic_.clear();
        n_ = rows_ * cols_;
    }

private:
    int rows_;
    int cols_;
    int n_;
    unordered_map<int, int> dic_;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(n_rows, n_cols);
 * vector<int> param_1 = obj->flip();
 * obj->reset();
 */

File: algorithms/python/randomFlipMatrix/randomFlipMatrix.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/random-flip-matrix/
# Author: Miao Zhang
# Date:   2021-02-19

class Solution:

    def __init__(self, n_rows: int, n_cols: int):
        self.rows = n_rows
        self.cols = n_cols
        self.n = n_rows * n_cols
        self.dic = dict()

    def flip(self) -> List[int]:
        s = random.randint(0, self.n - 1)
        self.n -= 1
        index = self.dic.get(s, s)
        self.dic[s] = self.dic.get(self.n, self.n)
        return [index // self.cols, index % self.cols]

    def reset(self) -> None:
        self.dic.clear()
        self.n = self.rows * self.cols


# Your Solution object will be instantiated and called as such:
# obj = Solution(n_rows, n_cols)
# param_1 = obj.flip()
# obj.reset()
